File: BackEnd/src/data/IMappers/INotificationMapper.ts
/**
 * @file INotificationMapper.ts
 * @description Interfaz del mapper de notificaciones in-app.
 * @module Reminders
 * @layer Data
 */

import type { Notification } from '../../domain/entities/Notification.js';
import type { NotificationResponseDTO } from '../../domain/dtos/notifications/notification-response.dto.js';
import type { NotificationDocument } from '../datasources/mongodb/models/NotificationModel.js';

export interface INotificationMapper {
  /**
   * Convierte un documento MongoDB en entidad de dominio.
   *
   * @param doc — Documento de la colección notifications.
   * @returns Entidad Notification.
   */
  toEntity(doc: NotificationDocument): Notification;

  /**
   * Convierte una entidad en documento listo para MongoDB.
   *
   * @param entity — Entidad Notification.
   * @returns NotificationDocument.
   */
  toDocument(entity: Notification): NotificationDocument;

  /**
   * Convierte una entidad en DTO de respuesta HTTP.
   *
   * @param entity — Entidad Notification.
   * @returns NotificationResponseDTO.
   */
  toResponseDTO(entity: Notification): NotificationResponseDTO;
}

File: FrontEnd/lib/data/i_mappers/i_notification_mapper.dart
/// @file i_notification_mapper.dart
/// @description Interfaz del mapper de notificaciones in-app.
/// Contrato NotificationModel ↔ AppNotification.
/// @module Reminders
/// @layer Data
library;

import '../../domain/entities/notification.dart';
import '../models/notification_model.dart';

/// Contrato de conversión entre el modelo de serialización y la entidad de dominio.
///
/// [injectable] registrar en container.dart como singleton.
abstract interface class INotificationMapper {
  /// Convierte un modelo de serialización a entidad de dominio.
  AppNotification toEntity(NotificationModel model);

  /// Convierte una entidad de dominio a modelo de serialización.
  NotificationModel toModel(AppNotification entity);
}
